lib/services/fanout.go

```diff
@@ -354,3 +359,87 @@ func (w *fanoutWatcher) Error() error {
 		return nil
 	}
 }
+
+// fanoutSetSize is the number of members in a fanout set.  selected based on some experimentation with
+// the FanoutSetRegistration benchmark.  This value keeps 100K concurrent registrations well under 1s.
+const fanoutSetSize = 128
+
+// FanoutSet is a collection of separate Fanout instances. It exposes an identical API, and "load balances"
+// watcher registration across the enclosed instances. In very large clusters it is possible for tens of
+// thousands of nodes to simultaneously request watchers. This can cause serious contention issues. FanoutSet is
+// a simple but effective solution to that problem.
+type FanoutSet struct {
+	// rw mutex is used to ensure that Close and Reset operations are exclusive,
+	// since these operations close watchers. Enforcing this property isn't strictly
+	// necessary, but it prevents a scenario where watchers might observe a reset/close,
+	// attempt re-registration, and observe the *same* reset/close again. This isn't
+	// necessarily a problem, but it might confuse attempts to debug other event-system
+	// issues, so we choose to avoid it.
+	rw      *sync.RWMutex
+	counter *atomic.Uint64
+	members []*Fanout
+}
+
+// NewFanoutSet creates a new FanoutSet instance in an uninitialized
+// state.  Until initialized, watchers will be queued but no
+// events will be sent.
+func NewFanoutSet() FanoutSet {
+	members := make([]*Fanout, 0, fanoutSetSize)
+	for i := 0; i < fanoutSetSize; i++ {
+		members = append(members, NewFanout())
+	}
+	return FanoutSet{
+		rw:      new(sync.RWMutex),
+		counter: atomic.NewUint64(0),
+		members: members,
+	}
+}
+
+// NewWatcher attaches a new watcher to a fanout instance.
+func (s FanoutSet) NewWatcher(ctx context.Context, watch types.Watch) (types.Watcher, error) {
+	s.rw.RLock()
+	defer s.rw.RUnlock()
+	fi := int(s.counter.Inc() % uint64(len(s.members)))
+	return s.members[fi].NewWatcher(ctx, watch)
+}
+
+// SetInit sets the Fanout instances into an initialized state, sending OpInit
+// events to any watchers which were added prior to initialization.
+func (s FanoutSet) SetInit() {
+	s.rw.RLock()
```

Comment:
Should we be holding a write lock here? Please leave a comment if not.

Reply:
Covered by comments on the `rw` field itself (takes a bit of explaining, so I decided to document on the field level rather than each individual call).

Reply:
Right. Maybe just leave a nod to it?

```suggestion
	// Write lock not needed, Fanout instances guard themselves.
	// See also the comment on FanoutSet.rw.
	s.rw.RLock()
```

Reply:
Although, to be fair, I'd take the write lock anyway - seems the correct thing to do and I'd hazard that it won't cause performance problems.